helix-lock/src/main/java/org/apache/helix/lock/helix/ZKDistributedNonblockingLock.java

```diff
@@ -144,19 +314,91 @@ public ZNRecord update(ZNRecord current) {
       // If no one owns the lock, allow the update
       // If the user is the current lock owner, allow the update
       LockInfo curLockInfo = new LockInfo(current);
-      if (!(System.currentTimeMillis() < curLockInfo.getTimeout()) || isCurrentOwner()) {
+      if (System.currentTimeMillis() > curLockInfo.getTimeout() || isCurrentOwner(curLockInfo)) {
         return _record;
       }
-      // For users who are not the lock owner and try to do an update on a lock that is held by
-      // someone else, exception thrown is to be caught by data accessor, and return false for
-      // the update
+
+      // higher priority lock request will preempt current lock owner that is with lower priority
+      if (!isCurrentOwner(curLockInfo) && _priority > curLockInfo.getPriority() && curLockInfo
+          .getRequestorId().equals("NONE")) {
+        // update lock Znode with requestor information
+        ZNRecord newRecord = composeNewRequestorRecord(curLockInfo, _record);
+        _isPending = true;
```

Comment:
As mentioned above, I think this would be better to set in the notification method.

Reply:
There is no notification that would be received by the requestor before it gets the lock. So we cannot se the pending state for it. The notification it receives is that the previous owner finished the cleanup work and it acquires the lock.

Reply:
Isn't it receiving notification on preempting?

Reply:
Notification would be received by all listeners. But for requestor, it won't handle its own preempting request. Only owner needs to start cleanup. Requestor only handle data change when owner finishes the cleanup work and remove itself from the owner field.

Reply:
I see. With the enum status instead of booleans, the logic has been clean enough. So I think it is good enough.